packages/commonwealth/server/util/tokenBalanceCache/providers/get_cw721_balances.ts

```diff
@@ -0,0 +1,131 @@
+import {
+  CosmWasmClient,
+  WasmExtension,
+  setupWasmExtension,
+} from '@cosmjs/cosmwasm-stargate';
+
+import { QueryClient } from '@cosmjs/stargate';
+import { HttpBatchClient, Tendermint34Client } from '@cosmjs/tendermint-rpc';
+import { factory, formatFilename } from 'common-common/src/logging';
+import { ChainNodeInstance } from '../../../models/chain_node';
+import { rollbar } from '../../rollbar';
+import { Balances } from '../types';
+
+const log = factory.getLogger(formatFilename(__filename));
+
+export type GetCw721BalancesOptions = {
+  chainNode: ChainNodeInstance;
+  addresses: string[];
+  contractAddress: string;
+  batchSize?: number;
+};
+
+export async function __getCw721Balances(
+  options: GetCw721BalancesOptions,
+): Promise<Balances> {
+  const { chainNode, contractAddress } = options;
+  let tmClient: Tendermint34Client;
+  if (options.addresses.length > 1) {
+    const batchClient = new HttpBatchClient(
+      chainNode.private_url || chainNode.url,
+      {
+        batchSizeLimit: options.batchSize || 100,
+        dispatchInterval: 10,
+      },
+    );
+    tmClient = await Tendermint34Client.create(batchClient);
+  } else {
+    tmClient = await Tendermint34Client.connect(
```

Comment:
done

Reply:
Edit: I actually found that `Tendermint34Client.connect` works better for single addresses. So I am keeping it here. I can't figure out why this is, but it is the pattern we use for get_cosmos_balances.

Reply:
>This would mean we can also extract the entire tmClient creation logic into a separate function since it is also used by the native cosmos token function.

done

Reply:
Using batch client for all cases after all. I believe the problems I was seeing before were related to handling multi vs single addresses.